lib/SPIRV/libSPIRV/SPIRVBasicBlock.cpp

```diff
@@ -87,6 +87,17 @@ void SPIRVBasicBlock::encodeChildren(spv_ostream &O) const {
 
 _SPIRV_IMP_ENCDEC1(SPIRVBasicBlock, Id)
 
+SPIRVInstruction *SPIRVBasicBlock::getVariableInsertionPoint() const {
```

Comment:
If I understand the code in SPIRVBasicBlock.cpp/h correctly, this function will set insert point in the beginning on the current basic block, but not the first basic block of the function. Regardless if I'm right or not, may I ask you to add a test showing, that even if alloca is in the second basic block, then OpVariable will still be added in the first one?

Reply:
This function will return the instruction corresponding to the first instruction not being `OpVariable`, `OpLine` or `OpNoLine`, i.e., where the next `OpVariable` should be inserted. In the case you mention, insertion will happen at that given point of the second block. Moving it to the first block may change original LLVM IR semantics, so I'd rather keep generating invalid code in that case as the input LLVM IR would be ill formed from a SPIR-V perspective. See the following case:

```llvm
loop_body:
  %ptr = alloca i32
  store ptr %ptr, ptr %ptrs
  ...
```

Representing the `alloca` instruction there as a single `OpVariable` would change semantics. Generating invalid SPIR-V as the original code is not representable as is in SPIR-V is what we were doing before and what we will keep doing after this patch. I do not intend to fix the general case, just the cases in which the `alloca` instructions are in the first block of a function, but not at the beginning of it.

Reply:
> Moving it to the first block may change original LLVM IR semantics

how actually ? the memory is uninitialised, the only thing to check is to make sure the alloca is not in a statically unreachable BB.

Reply:
> Representing the alloca instruction there as a single OpVariable would change semantics

IMHO no, it will only break instruction-level parallelism unless there is an IR analysis taking in account such cases (and I believe there should be something like this to unlock ILP in cases like this one https://godbolt.org/z/7c89dzra1

I'm not insisting to fix this case, the patch is already is a big improvement, just asking if we can/want to fix it now as well (I'm unsure what would be the effort)

Reply:
I was thinking of something like: https://godbolt.org/z/zfe1oq8nx. Moving the variable to the first block would effectively change the semantics to something like: https://godbolt.org/z/heK7To9P7.

Notice the original LLVM module in the first example, with an `alloca` instruction not in the first block, is already not representable in SPIR-V.